src/natcap/invest/recreation/recmodel_server.py

```diff
@@ -540,59 +827,96 @@ def _file_len(file_path):
+        n_workers=None, build_shapefile=True, fast_point_count=False):
     """Construct a spatial quadtree for fast querying of userday points.
 
     Args:
         initial_bounding_box (list of int):
-        raw_photo_csv_table ():
+        raw_csv_file_list (list): list of filepaths of point CSVs
+        dataset_name (string): one of 'flickr', 'twitter', indicating the
+            expected structure of the csv.
         cache_dir (string): path to a directory that can be used to cache
             the quadtree files on disk
+        ooc_qt_picklefilename (string): name for the pickle file quadtree index
+            created in the cache_dir.
         max_points_per_node(int): maximum number of points to allow per node
-            of the quadree.  A larger amount will cause the quadtree to
+            of the quadtree.  A larger amount will cause the quadtree to
             subdivide.
+        max_depth (int): maximum depth of nodes in the quadtree.
+            Once reached, the leaf nodes will not subdivide,
+            even if max_points_per_node is exceeded.
+        n_workers (int): number of cores for multiprocessing.
+        build_shapefile (boolean): whether or not to create vector geometries
+            representing nodes of the quadtree.
+        fast_point_count (boolean): If False, count the number of lines in all
+            the csv files. If True, estimate total number of points by counting
+            a random sample of files.
 
     Returns:
         None
     """
-    LOGGER.info('hashing input file')
-    start_time = time.time()
-    LOGGER.info(raw_photo_csv_table)
-    csv_hash = _hashfile(raw_photo_csv_table, fast_hash=True)
-
-    ooc_qt_picklefilename = os.path.join(cache_dir, csv_hash + '.pickle')
-    if os.path.isfile(ooc_qt_picklefilename):
-        return ooc_qt_picklefilename
+    LOGGER.info('counting lines in input file')
+    total_lines = _file_len(raw_csv_file_list, estimate=fast_point_count)
+    LOGGER.info('%d lines', total_lines)
+
+    # On a single CPU, flushing to disk is the main bottleneck.
+    # For large trees (i.e. twitter) more than 80% of wall-time
+    # is spent in the flush, while the main process is idle.
+    # Devoting 75% of CPUs to flush, setting 1 aside for the
+    # main process, leaving the rest for parsing input tables.
+    # When only 50% of CPUs devoted to flush, it is still a
+    # bottleneck and overall CPU efficiency according to SLURM
+    # is 25%, given 8 CPU. Many parser processes are idle most
+    # of the time also, as the numpy_array_queue is often full
+    if n_workers is None:
+        n_workers = multiprocessing.cpu_count()
+    n_flush_processes = int(n_workers * 0.75)
+    n_parse_processes = n_workers - n_flush_processes - 1
+    if n_parse_processes < 1:
+        n_parse_processes = 1
+    if n_flush_processes < 1:
+        n_flush_processes = 1
+
+    ooc_qt = out_of_core_quadtree.OutOfCoreQuadTree(
```

Comment:
Maybe a silly question, but what is the actual meaning of "out of core" for a quad tree?

Reply:
I remember asking Rich the same thing years ago. "out of core _memory_". I guess its a term for algorithms that use data too large for memory and are designed to use the disk as an alternative. When I talk about it I sometimes call it a "filesystem-based" quadtree.